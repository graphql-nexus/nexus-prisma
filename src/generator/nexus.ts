import { DMMF } from '@prisma/client/runtime'
import endent from 'endent'
import * as fs from 'fs-jetpack'
import * as OS from 'os'
import { GeneratedModule } from './types'

/** Transform the given DMMF into TS source code. */
export function generateRuntimeNexusSource(dmmf: DMMF.Document): GeneratedModule {
  if (process.env.NP_DEBUG) {
    fs.write('dmmf.json', dmmf)
  }

  const content = endent`
    import * as Nexus from 'nexus'

    ${dmmf.datamodel.models.map(modelToTypeScriptSource).join(OS.EOL + OS.EOL)}
  `

  return {
    fileName: 'nexus.ts',
    content,
  }
}

/** Transform a DMMF model into TS source code. */
function modelToTypeScriptSource(model: DMMF.Model): string {
  return endent`

    export const ${model.name} = {
      ${model.fields.filter(isSupportedType).map(fieldToTypeScriptSource).join(`,${OS.EOL}`)},
    } as const
  `
}

/**
 * Transform a DMMF model _field_ into TS source code.
 *
 * @example
 *   // Example of field data given:
 *   const field = {
 *     name: 'f16',
 *     kind: 'scalar',
 *     isList: false,
 *     isRequired: true,
 *     isUnique: false,
 *     isId: false,
 *     isReadOnly: false,
 *     type: 'DateTime',
 *     hasDefaultValue: false,
 *     isGenerated: false,
 *     isUpdatedAt: false,
 *   }
 */
function fieldToTypeScriptSource(field: DMMF.Field) {
  return endent`
    ${field.name}: {
      name: '${field.name}',
      config: {
        type: ${renderType(field)},
      }
    }
  `
}

function renderType(field: DMMF.Field): string {
  const graphqlType = getGraphQLType(field)

  if (field.isList) {
    return endent`
      Nexus.nonNull(Nexus.list(Nexus.nonNull(${quote(graphqlType)})))
    `
  } else if (field.isRequired) {
    return endent`
      Nexus.nonNull(${quote(graphqlType)})
    `
  } else {
    return endent`
      Nexus.nullable(${quote(graphqlType)})
    `
  }
}

type PrismaScalarType =
  | 'String'
  | 'Boolean'
  | 'Int'
  | 'BigInt'
  | 'Float'
  | 'Decimal'
  | 'DateTime'
  | 'Json'
  | 'Bytes'

function quote(x: string): string {
  return `'${x}'`
}

type GraphQLScalarType = 'ID' | 'String' | 'Int' | 'Float' | 'Boolean'

function isSupportedType(field: DMMF.Field): boolean {
  return field.kind === 'scalar'
}

function getGraphQLType(field: DMMF.Field): GraphQLScalarType {
  if (field.kind !== 'scalar') {
    throw new Error(`type mapping only supported for scalar-type fields`)
  }

  const typeName = field.type as PrismaScalarType

  switch (typeName) {
    case 'String': {
      if (field.isId) {
        return GraphQLScalarTypes.ID
      }
      return GraphQLScalarTypes.String
    }
    default: {
      return GraphQLScalarTypes.String
    }
  }
}

const GraphQLScalarTypes = {
  ID: 'ID',
  String: 'String',
} as const
