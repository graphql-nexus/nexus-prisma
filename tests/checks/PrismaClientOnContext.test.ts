import gql from 'graphql-tag'
import { objectType, queryType } from 'nexus'
import { testIntegration, TestIntegrationParams, testIntegrationPartial } from '../__helpers__/testers'

const base = testIntegrationPartial({
  datasourceSchema: `
    model User {
      id        String   @id
			profile   Profile? @relation(fields: [profileId], references: [id])
			profileId String
    }
		model Profile {
      id    String  @id
			user  User
		}
  `,
  apiSchema({ User, Profile }) {
    return [
      queryType({
        definition(t) {
          t.nonNull.list.nonNull.field('users', {
            type: 'User',
            resolve(_, __, ctx) {
              return [{ id: '1' }]
            },
          })
        },
      }),
      objectType({
        name: User.$name,
        definition(t) {
          t.field(User.id)
          t.field(User.profile)
        },
      }),
      objectType({
        name: Profile.$name,
        definition(t) {
          t.field(Profile.id)
        },
      }),
    ]
  },
  apiClientQuery: gql`
    query {
      users {
        profile {
          id
        }
      }
    }
  `,
})

/**
 * Break the instanceof check but duck typing will succeed.
 */
const instanceOfFailingPrismaClient: TestIntegrationParams['prismaClient'] = (prismaClientPackage) => {
  const prisma = new prismaClientPackage.PrismaClient()
  return { ...prisma }
}

/**
 * Some value that won't even pass Prisma Client duck typing check.
 */
const duckTypeFailingPrismaClient: TestIntegrationParams['prismaClient'] = (prismaClientPackage) => {
  return 'should be prisma client instance but is not' as any
}

testIntegration({
  description: 'error thrown when both instanceof check and duck typing fails',
  prismaClient: duckTypeFailingPrismaClient,
  ...base,
})

testIntegration({
  description: 'warning emitted when instanceof check fails but duck typing succeeds',
  prismaClient: instanceOfFailingPrismaClient,
  ...base,
})

testIntegration({
  description:
    'when check is disabled then warning NOT emitted when instanceof check fails but duck typing succeeds',
  prismaClient: instanceOfFailingPrismaClient,
  runtimeConfig(settings) {
    settings.change({
      checks: {
        PrismaClientOnContext: {
          enabled: false,
        },
      },
    })
  },
  ...base,
})
