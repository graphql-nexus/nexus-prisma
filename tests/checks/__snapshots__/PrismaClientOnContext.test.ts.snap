// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`If prisma client import fails then the check cannot perform its instanceof strategy and thus throws an error: graphqlOperationExecutionResult 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "profile": null,
      },
    ],
  },
  "errors": Array [
    [GraphQLError: Could not perform "PrismaClientOnContext" check because there was an error while trying to import Prisma Client:

Error: Cannot find module 'does-not-exist' from 'src/generator/models/javascript.ts'],
  ],
}
`;

exports[`If prisma client import fails then the check cannot perform its instanceof strategy and thus throws an error: graphqlSchemaSDL 1`] = `
"
type Query {
  users: [User!]!
}

type User {
  id: ID!
  profile: Profile
}

type Profile {
  id: ID!
}
"
`;

exports[`If prisma client import fails then the check cannot perform its instanceof strategy and thus throws an error: logs 1`] = `Array []`;

exports[`error thrown when both instanceof check and duck typing fails: graphqlOperationExecutionResult 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "profile": null,
      },
    ],
  },
  "errors": Array [
    [GraphQLError: Check "PrismaClientOnContext" failed. The GraphQL context.prisma value is not an instance of the Prisma Client.],
  ],
}
`;

exports[`error thrown when both instanceof check and duck typing fails: graphqlSchemaSDL 1`] = `
"
type Query {
  users: [User!]!
}

type User {
  id: ID!
  profile: Profile
}

type Profile {
  id: ID!
}
"
`;

exports[`error thrown when both instanceof check and duck typing fails: logs 1`] = `Array []`;

exports[`warning emitted when instanceof check fails but duck typing succeeds: graphqlOperationExecutionResult 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "profile": null,
      },
    ],
  },
}
`;

exports[`warning emitted when instanceof check fails but duck typing succeeds: graphqlSchemaSDL 1`] = `
"
type Query {
  users: [User!]!
}

type User {
  id: ID!
  profile: Profile
}

type Profile {
  id: ID!
}
"
`;

exports[`warning emitted when instanceof check fails but duck typing succeeds: logs 1`] = `
Array [
  "WARNING: Prisma Client on GraphQL context failed being checked using instanceof

REASON: The Prisma Client class reference imported from /home/debian/projects/prisma/nexus-prisma/tests/__cache__/integration/warning-emitted-when-instanceof-check-fails-but-duck-typing-succeeds/client is not the same class used by you to create your Prisma Client instance.

CONSEQUENCE: Maybe none since duck typing fallback indicates that the Prisma Client on the GraphQL context is actually valid. However relying on duck typing is hacky.

CODE: PrismaClientOnContextInstanceOfStrategyFailed",
]
`;

exports[`when check instanceof strategy is disabled then warning NOT emitted when instanceof check fails but duck typing succeeds: graphqlOperationExecutionResult 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "profile": null,
      },
    ],
  },
}
`;

exports[`when check instanceof strategy is disabled then warning NOT emitted when instanceof check fails but duck typing succeeds: graphqlSchemaSDL 1`] = `
"
type Query {
  users: [User!]!
}

type User {
  id: ID!
  profile: Profile
}

type Profile {
  id: ID!
}
"
`;

exports[`when check instanceof strategy is disabled then warning NOT emitted when instanceof check fails but duck typing succeeds: logs 1`] = `Array []`;

exports[`when check is disabled then warning NOT emitted when instanceof check fails OR duck typing fails: graphqlOperationExecutionResult 1`] = `
Object {
  "data": Object {
    "users": Array [
      Object {
        "profile": null,
      },
    ],
  },
  "errors": Array [
    [GraphQLError: Cannot read properties of undefined (reading 'findUnique')],
  ],
}
`;

exports[`when check is disabled then warning NOT emitted when instanceof check fails OR duck typing fails: graphqlSchemaSDL 1`] = `
"
type Query {
  users: [User!]!
}

type User {
  id: ID!
  profile: Profile
}

type Profile {
  id: ID!
}
"
`;

exports[`when check is disabled then warning NOT emitted when instanceof check fails OR duck typing fails: logs 1`] = `Array []`;
